cl/persistence/beacon_indexes/sql_indexer.go

```diff
@@ -0,0 +1,138 @@
+package beacon_indexes
+
+import (
+	"database/sql"
+	"fmt"
+
+	libcommon "github.com/ledgerwatch/erigon-lib/common"
+	"github.com/ledgerwatch/erigon/cl/cltypes"
+	_ "github.com/mattn/go-sqlite3"
+)
+
+// Define all queries templates
+var beaconIndiciesCreateTableQuery = `CREATE TABLE beacon_indicies (
+    Slot INTEGER NOT NULL,
+	BeaconBlockRoot BLOB NOT NULL CHECK(length(BeaconBlockRoot) = 32), -- Ensure it's 32 bytes
+	StateRoot BLOB NOT NULL CHECK(length(BeaconBlockRoot) = 32),
+	ParentBlockRoot BLOB NOT NULL CHECK(length(BeaconBlockRoot) = 32),
+    Canonical INTEGER NOT NULL DEFAULT 0, -- 0 for false, 1 for true
+    PRIMARY KEY (Slot, BeaconBlockRoot)  -- Composite key ensuring unique combination of Slot and BeaconBlockRoot
+);`
+
+var beaconIndiciesUniqueIndiciesQuery = `
+	CREATE UNIQUE INDEX idx_unique_canonical 
+	ON beacon_indicies (Slot) 
+	WHERE Canonical = 1;`
+
+var beaconReadBlockSlotByBlockRoot = "SELECT Slot FROM beacon_indicies WHERE BeaconBlockRoot = ?"
+
+var beaconReadParentRootByBlockRoot = "SELECT ParentBlockRoot FROM beacon_indicies WHERE BeaconBlockRoot = ?"
+
+var beaconReadCanonicalBlockRootBySlot = `SELECT BeaconBlockRoot FROM beacon_indicies 
+	WHERE Slot = ? AND Canonical = 1`
+
+var markRootCanonicalQuery = `UPDATE beacon_indicies 
+	SET Canonical = 1 
+	WHERE Slot = ? AND BeaconBlockRoot = ?`
+
+var resetOtherRootsCanonicalQuery = `UPDATE beacon_indicies 
+	SET Canonical = 0 
+	WHERE Slot = ? AND BeaconBlockRoot != ?`
+
+var insertOrUpdateBlockRoot = `
+	INSERT INTO beacon_indicies (Slot, BeaconBlockRoot, StateRoot, ParentBlockRoot, Canonical) 
+	VALUES (?, ?, ?, ?, 0);
+`
+
+type SqliteBeaconIndexer struct {
+	db *sql.DB
+}
+
+func NewSqlBeaconIndexer(db *sql.DB) (SqliteBeaconIndexer, error) {
+	if _, err := db.Exec(beaconIndiciesCreateTableQuery); err != nil {
+		return SqliteBeaconIndexer{}, err
+	}
+	if _, err := db.Exec(beaconIndiciesUniqueIndiciesQuery); err != nil {
+		return SqliteBeaconIndexer{}, err
+	}
+	// Initialize the table for indicies
+	return SqliteBeaconIndexer{
+		db: db,
+	}, nil
+}
+
+func (s SqliteBeaconIndexer) ReadBlockSlotByBlockRoot(blockRoot libcommon.Hash) (uint64, error) {
+	var slot uint64
+
+	// Execute the query.
+	err := s.db.QueryRow(beaconReadBlockSlotByBlockRoot, blockRoot[:]).Scan(&slot) // Note: blockRoot[:] converts [32]byte to []byte
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return 0, fmt.Errorf("no block found with the provided BeaconBlockRoot: %v", err)
+		}
+		return 0, fmt.Errorf("failed to retrieve slot for BeaconBlockRoot: %v", err)
+	}
+
+	return slot, nil
```

Comment:
instead of declaring an extra variable on the stack to cast into, just use named return args
```
func (s SqliteBeaconIndexer) ReadBlockSlotByBlockRoot(blockRoot libcommon.Hash) (slot uint64, err error) {
}
```